File: 13.罗马数字转整数.cs
/*
 * @lc app=leetcode.cn id=13 lang=csharp
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
public class Solution
{
    static Dictionary<char, int> data = new Dictionary<char, int>{
        {'I', 1},
        {'V', 5},
        {'X', 10},
        {'L', 50},
        {'C', 100},
        {'D', 500},
        {'M', 1000}
    };

    public int RomanToInt(string s)
    {
        int length = s.Length;
        var source = s.ToCharArray();

        int result = 0;
        string unit = "";

        for (int i = 0; i < length; i++)
        {
            unit += source[i];

            if (i < length - 1 && data[source[i]] < data[source[i + 1]])
            {
                continue;
            }

            // 取出unit的值
            var temp = unit.ToCharArray();
            if (temp.Length == 1)
            {
                result += data[temp[0]];
            }
            else
            {
                for (int j = 0; j < temp.Length; j++)
                {
                    if (j != temp.Length - 1)
                    {
                        result -= data[temp[j]];
                    }
                    else
                    {
                        result += data[temp[j]];
                    }
                }
            }

            unit = "";
        }

        return result;
    }
}
// @lc code=end

File: 13.罗马数字转整数.go
/*
 * @lc app=leetcode.cn id=13 lang=golang
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
func romanToInt(s string) int {
	base = make(map[byte]int)
	base["I"[0]] = 1
	base["V"[0]] = 5
	base["X"[0]] = 10
	base["L"[0]] = 50
	base["C"[0]] = 100
	base["D"[0]] = 500
	base["L"[0]] = 1000

	len := len(s)
	result := 0

	for i, v := range s {
		unit := string(v)

		if int(i+1) < len {
			j := i

			for base[s[j]] < base[s[j+1]] || (base[s[j]] == base[s[j+1]] && s[j] == "I"[0]) {
				unit = unit + string(s[j+1])
			}
		}

		result = transPart(unit, result)
	}

	return result
}

var base map[byte]int

func transPart(s string, origin int) int {
	if len(s) == 1 {
		return origin*10 + base[s[0]]
	} else if s[len(s)-1] == "I"[0] {
		return origin + 10 + len(s)
	} else {
		return origin*10 + (base[s[1]] - base[s[0]])
	}
}

// @lc code=end

